### PR TITLE
Add support for test filtering via tags

### PR DESCRIPTION
**Pull request by yannbf**
Closes: #281, Closes: #218, Closes: #74, Closes: #306, Closes: #247

This PR introduces the concept of test filtering via the story tags annotation.

## Filtering tests (experimental)

You might want to skip certain stories in the test-runner, run tests only against a subset of stories, or exclude certain stories entirely from your tests. This is possible via the `tags` annotation.

This annotation can be part of a story, therefore only applying to it, or the component meta (the default export), which applies to all stories in the file:

```ts
const meta = {
  component: Button,
  tags: ['design', 'test-only'],
};
export default meta;

// will inherit tags from meta with value ['design', 'test-only']
export const Primary = {};

export const Secondary = {
  // will override tags to be just ['skip']
  tags: ['skip'],
};
```

> **Note**
> You can't import constants from another file and use them to define tags in your stories. The tags in your stories or meta **have to be** defined inline, as an array of strings. This is a limitation due to Storybook's static analysis.

Once your stories have your own custom tags, you can filter them via the tags property in your test-runner configuration file. You can also use the CLI flags `--includeTags`, `--excludeTags` or `--skipTags` for the same purpose. The CLI flags will take precedence over the tags in the test-runner config, therefore overriding them.

#### Test-runner config: tags

The `tags` property contains three options: `include | exclude | skip`, each accepting an array of strings:

```js
// .storybook/test-runner.js
module.exports = {
  tags: {
    include: [], // string array, e.g. ['test-only']
    exclude: [], // string array, e.g. ['design', 'docs-only']
    skip: [], // string array, e.g. ['design']
  },
};
```

<details>
  <summary>📦 Published PR as canary version: <code>0.15.0--canary.382.449be19.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/test-runner@0.15.0--canary.382.449be19.0
  # or 
  yarn add @storybook/test-runner@0.15.0--canary.382.449be19.0
  ```
</details>

### CODE REVIEW

File: src/playwright/transformPlaywrightJson.test.ts [hunk 1]
```diff
@@ -198,6 +207,165 @@ describe('Playwright Json', () => {
...
+            importPath: './stories/basic/Page.stories.js',
+          },
+        },
+      };
+      // Should result in:
+      // - A being excluded
+      // - B being included, but skipped
+      // - C being included
+      // - D being excluded
+      expect(transformPlaywrightJson(input)).toMatchInlineSnapshot(`
```
Comment (ndelangen) [inline src/playwright/transformPlaywrightJson.test.ts:251]: I'm not a huge fan of this huge snapshot tbh.
splitting this into 2 files: input & output and then asserting with given input, the output matches the file would make it easier to manually check if the output doesn't contain some small mistake.

There's 0 syntax safety here.

**Review by ndelangen — APPROVED** [review]

File: src/csf/transformCsf.ts [hunk 2]
```diff
@@ -91,23 +114,49 @@ export const transformCsf = (
...
 
-  const storyPlays = storyExports.reduce((acc, key) => {
+  const storyAnnotations = storyExports.reduce((acc, key) => {
     const annotations = csf._storyAnnotations[key];
+    acc[key] = {};
     if (annotations?.play) {
-      acc[key] = annotations.play;
+      acc[key].play = annotations.play;
     }
+    acc[key].tags = annotations.tags ? parseTags(annotations.tags) : csf.meta.tags || [];
```
Comment (shilman) [inline src/csf/transformCsf.ts:131]: Tags should be available on `csf._stories[key].tags`

**Review by shilman — APPROVED** [review]
Looks great! Minor comment on the code. Also, you might want to provide a recipe for how to only execute tests with a play function using tags, since that's a frequently requested feature.

File: README.md [hunk 3]
```diff
@@ -159,6 +161,9 @@ Usage: test-storybook [options]
 | `--ci`                            | Instead of the regular behavior of storing a new snapshot automatically, it will fail the test and require Jest to be run with `--updateSnapshot`. <br/>`test-storybook --ci` |
 | `--shard [shardIndex/shardCount]` | Splits your test suite across different machines to run in CI. <br/>`test-storybook --shard=1/3`                                                                              |
 | `--failOnConsole`                 | Makes tests fail on browser console errors<br/>`test-storybook --failOnConsole`                                                                                               |
+| `--includeTags`                   | Only test stories that match the specified tags, comma separated<br/>`test-storybook --includeTags="test-only"`                                                               |
```
Comment (jonniebigodes) [inline README.md:164]: @yannbf one question for you, shouldn't these be tagged as experimental matching the filtering section that was added? Don't won't introduce confusion to anyone who is reading this, and start asking questions about it. Let me know what you think and we'll take it from there.

Comment (yannbf) [reply README.md:164]: how would you suggest this change be done?

**Comment by IanVS** [comment]
Not sure where the best place to ask this is, but what is the difference between `skipTags` and `excludeTags`?  From reading the description, both of those will cause stories with those tags not to execute in the test runner, right?  Why have two different tags that do the same thing?

**Comment by yannbf** [comment]
> Not sure where the best place to ask this is, but what is the difference between `skipTags` and `excludeTags`? From reading the description, both of those will cause stories with those tags not to execute in the test runner, right? Why have two different options that do the same (?) thing?
> 
> Edit: I see, `skipTags` still shows the story in the cli output:
> 
> image
> vs `excludeTags` doesn't:
> 
> image
> Personally I don't see a compelling reason to support both options, and probably would have just started with `excludeTags` for now until/unless people asked for a way to also list them as skipped. But since both now exist, I'd suggest we need to clarify the difference between them (maybe "Mark stories as skipped" vs "Skip test stories").

Thanks for your input Ian! And thanks for working on the clarification in the docs. Skipping is quite useful when there are temporary issues that you want to disable tests for, but you don't want to forget that they exist

**Pull request merged**